src/extra_files.c

```diff
@@ -0,0 +1,165 @@
+
+static struct fileskip {
+	const char *shortname; /* Without prefix */
+	char *name;	    /* full name, including any path_prefix */
+	int len;
+} skippers[] =
+    { { shortname : "/lib/modules" },
+      { shortname : "/local" } };
+static int path_prefix_len;
+
+/* Helper function to call from nftw */
+
+static int record_filename(const char *name, const struct stat *stat __attribute__((unused)), int type, struct FTW *ftw __attribute__((unused)))
+{
+	for (size_t i = 0; i < sizeof(skippers) / sizeof(skippers[0]); i++) {
+		if (strncmp(name, skippers[i].name, skippers[i].len) == 0) {
+			return FTW_SKIP_SUBTREE;
+		}
+	}
+
+	char *savedname = strdup(name + path_prefix_len - 1); /* Only store name relative to top of area */
+	F = realloc(F, (nF + 1) * sizeof(*F));		      /* TODO, check realloc is smart, so don't need to double myself */
+	if (!F || !name) {
+		fprintf(stderr, "Out of memory allocating %d filenames\n", nF);
+		return -ENOMEM;
+	}
+	F[nF].filename = savedname;
+	F[nF].dir = (type == FTW_D);
+	F[nF].in_manifest = false; /* Because we do yet know */
+	nF++;
+	return 0;
+}
+
+/* qsort helper function */
+static int qsort_helper(const void *A, const void *B)
+{
+	return strcmp(((struct filerecord *)A)->filename, ((struct filerecord *)B)->filename);
+}
+/* bsearch helper function */
+static int looker(const void *A, const void *B)
+{
+	return strcmp(*(const char **)A, ((struct filerecord *)B)->filename);
+}
+
+/* expect the start to end in /usr and be the absolute path to the root */
+int walk_tree(struct manifest *manifest, const char *start)
+{
+	/* Walk the tree, */
+	int rc;
+	path_prefix_len = strlen(path_prefix);
+	/* Set up the directories to skip */
+	for (size_t i = 0; i < sizeof(skippers) / sizeof(skippers[0]); i++) {
+		skippers[i].name = mk_full_filename(start, skippers[i].shortname);
+		skippers[i].len = strlen(skippers[i].name);
+	}
+	rc = nftw(start, &record_filename, 0, FTW_ACTIONRETVAL | FTW_PHYS | FTW_MOUNT);
+	const char *skip_dir = NULL; /* Skip files below this in printout */
+	if (rc) {
+		goto tidy; /* Already printed out of memory */
+	}
+	qsort(F, nF, sizeof(*F), &qsort_helper);
+	/* Interesting question, would it be faster to sort this linked list,
+	 * or convert it to an array of pointers, or just pull them off one
+	 * at a time? Try one at a time first.
+	 */
+	struct list *iter = list_head(manifest->files);
+	while (iter) {
+		struct file *file;
+		struct filerecord *found;
+
+		file = iter->data;
+		iter = iter->next;
+
+		if (file->is_deleted) {
+			continue;
+		}
+		found = bsearch(&file->filename, F, nF, sizeof(*F), &looker);
+		if (found) {
+			found->in_manifest = true;
+		}
+	}
+	/* list files/directories which are extra */
+	for (int i = 0; i < nF; i++) {
+		int skip_len; /* Length of directory name we are skipping */
+		if (!F[i].in_manifest) {
+			if (skip_dir) {
+				/* This is under the current directory to be skipped */
+				if ((strncmp(F[i].filename, skip_dir, skip_len) == 0) &&
```

Comment:
It is not used before it is set in line 147. If skip_dir is NULL then skip_len is not used. I could set it to be 0 or 42 or whatever but it wouldn't every be used.